File: src/main/java/edu/cmu/isr/ExtTLASpec.java
package edu.cmu.isr;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 *
 */
public class ExtTLASpec {

  /**
   *
   */
  private Map<String, ExtTLAModule> modules = new HashMap<>();

  public ExtTLAModule addNewModule(String name) {
    ExtTLAModule module = new ExtTLAModule(name);
    modules.put(name, module);
    return module;
  }

  public ExtTLAModule extendModule(ExtTLAModule m) {
    List<ExtTLAModule> extModules = new LinkedList<>();
    m.getExtendModules().forEach(i -> {
      if (modules.containsKey(i)) {
        // Extend the module first
        extModules.add(extendModule(modules.get(i)));
      } else {
        throw new Error("no such module " + i);
      }
    });
    return m.extendWith(extModules);
  }

  public Map<String, ExtTLAModule> getModules() {
    return modules;
  }
}

File: src/main/java/edu/cmu/isr/exttla/ExtTLASpec.kt
package edu.cmu.isr.exttla

class ExtTLASpec {

  val modules: MutableMap<String, ExtTLAModule> = mutableMapOf()

  fun addNewModule(name: String): ExtTLAModule {
    val module = ExtTLAModule(name)
    modules[name] = module
    return module
  }

  fun extendModule(m: ExtTLAModule): ExtTLAModule {
    return m.extendWith(findParentModules(m))
  }

  private fun findParentModules(m: ExtTLAModule): MutableList<ExtTLAModule> {
    return m.extendModules
      .map {
        if (it in modules) {
          val parents = findParentModules(modules[it]!!)
          parents.add(modules[it]!!)
          parents
        } else {
          throw Error("no such module $it")
        }
      }.fold(mutableListOf()) { l, it ->
        it.forEach { mm ->
          if (!l.contains(mm)) {
            l.add(mm)
          }
        }
        l
      }
  }
}